File: Python/7kyu/Sum of digits - Digital Root.py
# https://www.codewars.com/kata/541c8630095125aba6000c00/train/python

def digital_root(n):
    # Recursive solution, return sum of all digits in n if < 9 else recursive call with that sum
    return sum([int(i) for i in str(n)]) if sum([int(i) for i in str(n)])<10 else digital_root(sum([int(i) for i in str(n)]))


if __name__ == "__main__":
    print(digital_root(16), 7)
    print(digital_root(942), 6)
    print(digital_root(132189), 6)
    print(digital_root(493193), 2)


# def digital_root(n):return n if n < 10 else digital_root(sum(map(int,str(n))))
# def digital_root(n):return n%9 or n and 9 
# def digital_root(n):return 1 + ((int(n) - 1) % 9) if n>0 else 0

File: C/7kyu/Sum of Digits - Digital Root.c
#include <stdio.h>
/*
https://www.codewars.com/kata/541c8630095125aba6000c00/train/c
*/

int digital_root(int n);


int main(){
    printf("Expected %d, got %d\n", 7, digital_root(16));
    printf("Expected %d, got %d\n", 6, digital_root(195));
    printf("Expected %d, got %d\n", 2, digital_root(992));
    printf("Expected %d, got %d\n", 9, digital_root(167346));
    printf("Expected %d, got %d\n", 0, digital_root(0));
    return 0;
}


int digital_root(int n){
    int sum = 0;

    while (n>0){
        sum += n%10;
        n/=10;
        if (n==0 && sum >= 10){
            n=sum;
            sum=0;
        }
    }
    return sum;
}


/*
int digital_root(int Z) {
  return --Z % 9 + 1;
}
*/

/*
int digital_root(int n) {
  return (n-1)%9 + 1;
}
*/

/*
int digital_root(int n) {
  return n ? ((n = n % 9) ? n : 9) : 0;
}
*/

/*

*/
